# Biodiversity in U.S. National Parks Portfolio Project

## Introduction

## Load and Inspect the Data

In [7]:
import pandas as pd
import glob

In [9]:
species = pd.read_csv('species_info.csv')
observations = pd.read_csv('observations.csv')

In [10]:
species.head()

,category,scientific_name,common_names,conservation_status
0,Mammal,Clethrionomys gapperi gapperi,Gapper's Red-Backed Vole,NaN
1,Mammal,Bos bison,"American Bison, Bison",NaN
2,Mammal,Bos taurus,"Aurochs, Aurochs, Domestic Cattle (Feral), Dom...",NaN
3,Mammal,Ovis aries,"Domestic Sheep, Mouflon, Red Sheep, Sheep (Feral)",NaN
4,Mammal,Cervus elaphus,Wapiti Or Elk,NaN


In [11]:
species.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5824 entries, 0 to 5823
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   category             5824 non-null   object
 1   scientific_name      5824 non-null   object
 2   common_names         5824 non-null   object
 3   conservation_status  191 non-null    object
dtypes: object(4)
memory usage: 182.1+ KB


In [12]:
observations.head()

,scientific_name,park_name,observations
0,Vicia benghalensis,Great Smoky Mountains National Park,68
1,Neovison vison,Great Smoky Mountains National Park,77
2,Prunus subcordata,Yosemite National Park,138
3,Abutilon theophrasti,Bryce National Park,84
4,Githopsis specularioides,Great Smoky Mountains National Park,85


In [13]:
observations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23296 entries, 0 to 23295
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   scientific_name  23296 non-null  object
 1   park_name        23296 non-null  object
 2   observations     23296 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 546.1+ KB


## Clean Data

### Check for and remove duplicates

In [16]:
species_duplicates = species.duplicated()
species_duplicates.value_counts()

False    5824
Name: count, dtype: int64

In [17]:
observations_duplicates = observations.duplicated()
observations_duplicates.value_counts()

False    23281
True        15
Name: count, dtype: int64

While the **species** dataframe contains no duplicates, 15 duplicates are found in the **observations** database. Let's go ahead and remove those duplicates.

In [18]:
observations = observations.drop_duplicates()
observations_duplicates = observations.duplicated()
observations_duplicates.value_counts()

False    23281
Name: count, dtype: int64

Now both dataframes are without duplicates.

### Check data types

In [21]:
species.dtypes

category               object
scientific_name        object
common_names           object
conservation_status    object
dtype: object

In [22]:
observations.dtypes

scientific_name    object
park_name          object
observations        int64
dtype: object

Both datasets have the proper data types assigned to their columns (i.e. a column with numeric information like *observations* in the **observations** dataframe has an *int64* data type, not an *object* data type).

### Fill missing values

Let's check whether any of the columns contain missing values. I already know that the *conservation_status* column in the **species** dataframe includes NaN values, but are there any others?

In [23]:
species.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5824 entries, 0 to 5823
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   category             5824 non-null   object
 1   scientific_name      5824 non-null   object
 2   common_names         5824 non-null   object
 3   conservation_status  191 non-null    object
dtypes: object(4)
memory usage: 182.1+ KB


In [24]:
observations.info()

<class 'pandas.core.frame.DataFrame'>
Index: 23281 entries, 0 to 23295
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   scientific_name  23281 non-null  object
 1   park_name        23281 non-null  object
 2   observations     23281 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 727.5+ KB


In the **species** dataframe, among the 5824 entries, only the aforementioned *conservation_status* column contains missing values. In the **observations** dataframe, among the 23281 entries, all three columns do not have any missing values. Therefore, I must only focus on the *conservation_status* column when dealing with missing values.

Because I want to know about the content of the **species** dataframe aside from the *conservation_status* column, I will replace any NaN values with "Unknown" (I know from checking the data types above that the *conservation_status* column features object values).

In [25]:
species.conservation_status = species.conservation_status.fillna("Unknown")
species.head()

,category,scientific_name,common_names,conservation_status
0,Mammal,Clethrionomys gapperi gapperi,Gapper's Red-Backed Vole,Unknown
1,Mammal,Bos bison,"American Bison, Bison",Unknown
2,Mammal,Bos taurus,"Aurochs, Aurochs, Domestic Cattle (Feral), Dom...",Unknown
3,Mammal,Ovis aries,"Domestic Sheep, Mouflon, Red Sheep, Sheep (Feral)",Unknown
4,Mammal,Cervus elaphus,Wapiti Or Elk,Unknown


In [26]:
species.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5824 entries, 0 to 5823
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   category             5824 non-null   object
 1   scientific_name      5824 non-null   object
 2   common_names         5824 non-null   object
 3   conservation_status  5824 non-null   object
dtypes: object(4)
memory usage: 182.1+ KB


Now that duplicates and null values have been removed from the datasets, I can now move forward with analysing the datasets.